src/language_features/hover.rs

```diff
@@ -49,28 +77,43 @@ pub fn editor_hover(
     cursor: KakounePosition,
     range: KakouneRange,
     tabstop: usize,
-    result: Option<Hover>,
+    results: Vec<(ServerName, Option<Hover>)>,
     ctx: &mut Context,
 ) {
     let doc = &ctx.documents[&meta.buffile];
-    let lsp_range = kakoune_range_to_lsp(&range, &doc.text, ctx.offset_encoding);
+    let lsp_ranges: HashMap<_, _> = results
+        .iter()
+        .map(|(server_name, _)| {
+            let offset_encoding = ctx.language_servers[server_name].offset_encoding;
+            (
+                server_name,
+                kakoune_range_to_lsp(&range, &doc.text, offset_encoding),
+            )
+        })
+        .collect();
     let for_hover_buffer = matches!(hover_type, HoverType::HoverBuffer { .. });
     let diagnostics = ctx.diagnostics.get(&meta.buffile);
     let diagnostics = diagnostics
         .map(|x| {
             x.iter()
-                .filter(|x| ranges_touch_same_line(x.range, lsp_range))
-                .filter(|x| !x.message.is_empty())
-                .map(|x| {
+                .filter(|(server_name, x)| {
+                    lsp_ranges
+                        .get(server_name)
+                        .filter(|lsp_range| ranges_touch_same_line(x.range, **lsp_range))
+                        .is_some()
+                })
+                .filter(|(_, x)| !x.message.is_empty())
+                .map(|(server_name, x)| {
+                    let server = &ctx.language_servers[server_name];
                     // Indent line breaks to the same level as the bullet point
                     let message = (x.message.trim().to_string()
-                        + &format_related_information(x, ctx)
+                        + &format_related_information(x, (server_name, server), server, ctx)
                             .map(|s| "\n  ".to_string() + &s)
                             .unwrap_or_default())
                         .replace('\n', "\n  ");
                     if for_hover_buffer {
                         // We are typically creating Markdown, so use a standard Markdown enumerator.
-                        return format!("* {}", message);
+                        return format!("* [{server_name}] {message}");
```

Comment:
this still shows "c_cpp" here which is not the server name I think.
Also if there is only one server (common case) we don't want to name it here

Reply:
Yeah, I can add logic to omit such labels when there's a single server.

As for capabilities, are you okay with the output being ordered? I thought it'd be more consistent and easier to handle.

Reply:
> As for capabilities, are you okay with the output being ordered? I thought it'd be more consistent and easier to handle.

you mean the `(serverA, serverB)` labels in lsp-capabilities output? Probably ideal to use the input order (or input ranking once we have that) but alphabetic order is okay too

Reply:
I mean the capabilities per se. Previously, they were arbitrarily ordered (or maybe there was an implicit order), but because I had to use their names as a map key, I made my life easier by making the output consistent by using a BTreeMap.

If you wish the original order, I think we'd need to use something like [`indexmap`](https://github.com/bluss/indexmap).

Reply:
it's fine to sort capabilities, the previous order seems pretty random